lib/slave/cracker.go

```diff
@@ -8,14 +8,20 @@ import (
 	"github.com/aaronang/cong-the-ripper/lib/slave/hasher"
 )
 
-func Execute(task lib.Task, successChan chan CrackerSuccess, failChan chan CrackerFail) {
-	bd := brutedict.New(&task)
+func Execute(task *task, successChan chan CrackerSuccess, failChan chan CrackerFail) {
+	bd := brutedict.New(&task.Task)
 	var hasher hasher.Hasher = new(hasher.Pbkdf2) // Can be swapped with other hashing algorithms
 
 	log.Println("[ Task", task.ID, "]", "Start cracker.Execute")
 	for {
 		if candidate := bd.Next(); candidate != nil {
-			hash := hasher.Hash(candidate, &task)
+			select {
+			case c := <-task.progressChan:
+				c <- string(candidate)
+			default:
+			}
+
+			hash := hasher.Hash(candidate, &task.Task)
```

Comment:
I agree that the empty `default` could be confusing, but it just need to fall through if the channel is empty. Also your proposed structure does not work for this logic. First we want to check if the brutedict has a new candidate. If yes, then it need to check if it need to update the progressChan with the current candidate.

I am open for improvements on this chunk of code, but your proposed structure does not work.